Master

tycho01 opened the pull request:
Hey, just found your project yesterday, and think it's pretty cool. :)

I felt it was too bad your implementation didn't leave the definitions intact, so I tried to see if I could make those stay.

I think I ended up changing so much you may not like it (comments, apibase test failing too), but I don't mind if you don't merge, I just thought it'd be polite to show you what I tried at least. :)

horiuchi commented on src/writeProcessor.ts:
```diff
@@ -1,13 +1,14 @@
 
     public indentChar = ' ';
-    public indentStep = 4;
+    public indentStep = 2;
```
I thought we should almost follow the official [Coding Guideline](https://github.com/Microsoft/TypeScript/wiki/Coding-guidelines).
So `Use 4 spaces per indentation.`

horiuchi commented on test/expected_file/schema.d.ts:
```diff
@@ -1,50 +1,54 @@
-            not?: this;
-        }
+  namespace draft_04 {
+    export type PositiveInteger = number;
```
👍

horiuchi commented on test/expected_file/schema.d.ts:
```diff
@@ -1,50 +1,54 @@
-        }
+  namespace draft_04 {
+    export type PositiveInteger = number;
+    export interface PositiveIntegerDefault0 {}
```
But this type is not good...

horiuchi commented:
@tycho01 Thank you for your modification!
Looks almost good to me! 😄 
When you did cleanup code and fixed few issue, I will merge this PR!

tycho01 commented:
Cool :), will try later when time allows.

On Mon, May 9, 2016 at 11:52 AM, Horiuchi_H <notifications@github.com>
wrote:

> @tycho01 <https://github.com/tycho01> Thank you for your modification!
> Looks almost good to me! 😄
> When you did cleanup code and fixed few issue, I will merge this PR!
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly or view it on GitHub
> <https://github.com/horiuchi/dtsgenerator/pull/43#issuecomment-217770996>
>

horiuchi commented:
And I fixed the typo: TypeDef **e** nition.
Please `rebase master`.

tycho01 commented:
Thanks for your comments! I tried to make things a bit better now:
* cleaned up debugging statements
* I made sure your multi-line comments are possible again. Still using single-line as default though, haven't figured out a good balance yet. Feel free to change as you like.
* fixed d.ts indentation back to 4 spaces. Tried adjusting my own code a bit too, but my Atom automatically does 2, so that made it a bit harder.
* Tried to slightly improve handling of edge-cases like the one you mentioned (didn't fully work as intended, it ended up resolving the reference to `number` instead of keeping it as a reference to `PositiveInteger`). Things definitely get complicated with `allOf` + `$ref` as in this example, and I felt it was also getting harder to do the logic without using `lodash-fp` at this point.

I have to admit the complexity starts exploding with cases like this, so I don't feel very confident I've covered all edge cases like that either... I don't think I know of a good solution, it's definitely getting hard. :(

horiuchi commented:
Okay, Thank you for your commit! 👍 
I'll merge this pull request, and I'll fix the rest issues.

tycho01 commented:
I was trying out the result now, and found out some things weren't really going right yet, since it hadn't added namespaces where needed yet. This commit should fix that.
I ran it on all the schemas again (except `apimeta`, which failed from the start), so you can see all the resulting changes.

horiuchi commented on test/expected_file/simple_example.d.ts:
```diff
@@ -19,10 +15,7 @@ declare namespace json_schema_org {
-    }> {
-    }
+        // Coordinates of the warehouse with the product
+        warehouseLocation?: json_schema_org.Geo.geo;
```
I think this type should be `Geo`.

tycho01 replied:
Thanks, looks like you're right... let me look into that.

tycho01 commented:
Thanks for spotting that mistake. It turned out that issue also occurred in `hyper-schema` once. This should fix those two. Other schemas remained unaffected by the change.

horiuchi commented on test/expected_file/hyper-schema.d.ts:
```diff
@@ -0,0 +1,72 @@
+            maxLength?: PositiveInteger;
+            minLength?: PositiveIntegerDefault0;
+            pattern?: string; // regex
+            additionalItems?: boolean | Schema;
```
I concern that return type should be `this` instead of `Schema`.

horiuchi commented on test/expected_file/news.d.ts:
```diff
@@ -1,15 +1,39 @@
+    export type Detail = string;
+    // article's image url
+    //  * example: http://pressroom.lexus.com/images/Lexus_logo_bug_2013_thumb_20140604144404730.jpg
+    export interface ImageUrl {}
```
This type should be equal to `string`.

horiuchi commented:
I cannot fix this pull request code.
If you need to merge this, please separate this to small changes. Or I re-implement this idea.

tycho01 commented:
If you'd prefer reimplementing, I'm fine with that. I'm only really a casual user of this. I'd also agree that you should have the right to keep your code-base in line with your coding style.
I've found this whole thing frustrating, I don't even know

horiuchi commented:
@tycho01 Thank you for your contributions.
I'm sure that this idea is so useful and helpful for other users.
But this code is broken exists unit tests for few cases.

Now I'm reimplementing to support reference to the nodes under `definitions` node. Also now my direction is to support that this can generate from open api (`swagger` generated) format. But I take a while to do it.

tycho01 commented:
Yeah, it's definitely hard...
Thank you for considering to include the OpenAPI spec as well.


On Sun, Jun 19, 2016 at 1:16 AM, Horiuchi_H <notifications@github.com>
wrote:

> @tycho01 <https://github.com/tycho01> Thank you for your contributions.
> I'm sure that this idea is so useful and helpful for other users.
> But this code is broken exists unit tests for few cases.
>
> Now I'm reimplementing to support reference to the nodes under definitions
> node. Also now my direction is to support that this can generate from open
> api (swagger generated) format. But I take a while to do it.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/horiuchi/dtsgenerator/pull/43#issuecomment-226953853>,
> or mute the thread
> <https://github.com/notifications/unsubscribe/AC6uxQPLOFcOmtrhdpj9vFBqW4Jx9y1eks5qNCgJgaJpZM4IZrIL>
> .
>

horiuchi commented:
I supported the sub node types of `defenitions` node by #62.
